File: patient-case-taking-platform/apps/api/app/schemas/document_review.py
"""Public contracts for source-adjacent document review."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.documents.review import ReviewAction, ReviewState


class ReviewQueueItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    document_id: UUID
    facility_id: UUID
    patient_id: UUID
    encounter_id: UUID
    document_class: str
    state: str
    version: int
    updated_at: datetime
    candidate_count: int
    unresolved_count: int


class ReviewCandidateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    candidate_id: UUID
    entity_type: str
    normalized_value: str | None
    unit: str | None
    source_page_artifact_id: UUID
    source_ocr_artifact_id: UUID | None
    source_region_id: UUID | None
    source_page_number: int
    parser_signal: str
    negated: bool
    temporality: str
    subject: str
    uncertainty: str | None
    document_statement: bool
    clinician_confirmed_current: bool
    review_state: ReviewState
    version: int
    source_page_width: int | None
    source_page_height: int | None
    source_bbox: tuple[int, int, int, int] | None
    source_polygon: tuple[tuple[int, int], ...] | None


class ReviewDocumentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    document_id: UUID
    facility_id: UUID
    patient_id: UUID
    encounter_id: UUID
    purpose: str
    document_class: str
    state: str
    version: int
    updated_at: datetime
    candidates: tuple[ReviewCandidateResponse, ...]
    pages: tuple["ReviewPageResponse", ...]


class ReviewPageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    page_number: int
    width: int
    height: int
    preprocessing_version: str


class CandidateDecisionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    action: ReviewAction
    expected_candidate_version: int = Field(gt=0)
    corrected_value: str | None = Field(default=None, max_length=512)
    corrected_unit: str | None = Field(default=None, max_length=32)
    reason_code: str | None = Field(default=None, max_length=64)
    source_verified: bool = False


class ReviewFinalizeRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_document_version: int = Field(gt=0)


class OpenManualReviewRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_document_version: int = Field(gt=0)


class ManualCandidateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    entity_type: str = Field(min_length=1, max_length=64)
    normalized_value: str = Field(min_length=1, max_length=2_000)
    unit: str | None = Field(default=None, max_length=32)
    source_page_artifact_id: UUID
    source_page_number: int = Field(gt=0)
    source_verified: bool


class PagePreviewResponse(BaseModel):
    url: str
    expires_in_seconds: int

File: patient-case-taking-platform/apps/clinician-console/src/lib/documentReview.ts
import { clinicalAccessToken } from "./clinicalSession";

export type ReviewState =
  | "unreviewed"
  | "accepted"
  | "corrected"
  | "rejected"
  | "unreadable"
  | "rescan_requested"
  | "deferred";

export type ReviewAction =
  | "accept"
  | "correct"
  | "reject"
  | "unreadable"
  | "request_rescan"
  | "defer";

export type ReviewQueueItem = {
  document_id: string;
  facility_id: string;
  patient_id: string;
  encounter_id: string;
  document_class: string;
  state: string;
  version: number;
  updated_at: string;
  candidate_count: number;
  unresolved_count: number;
};

export type ReviewCandidate = {
  candidate_id: string;
  entity_type: string;
  normalized_value: string | null;
  unit: string | null;
  source_page_artifact_id: string;
  source_ocr_artifact_id: string | null;
  source_region_id: string | null;
  source_page_number: number;
  parser_signal: string;
  negated: boolean;
  temporality: string;
  subject: string;
  uncertainty: string | null;
  document_statement: boolean;
  clinician_confirmed_current: boolean;
  review_state: ReviewState;
  version: number;
  source_page_width: number | null;
  source_page_height: number | null;
  source_bbox: [number, number, number, number] | null;
  source_polygon: [number, number][] | null;
};

export type ReviewDocument = {
  document_id: string;
  facility_id: string;
  patient_id: string;
  encounter_id: string;
  purpose: string;
  document_class: string;
  state: string;
  version: number;
  updated_at: string;
  candidates: ReviewCandidate[];
  pages: ReviewPage[];
};

export type ReviewPage = {
  id: string;
  page_number: number;
  width: number;
  height: number;
  preprocessing_version: string;
};

export class ReviewApiError extends Error {
  constructor(
    message: string,
    readonly status: number,
  ) {
    super(message);
  }
}

const apiUrl = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";


async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${apiUrl}/api/v1/document-reviews${path}`, {
    ...init,
    cache: "no-store",
    headers: {
      Authorization: `Bearer ${await clinicalAccessToken()}`,
      "Content-Type": "application/json",
      ...init?.headers,
    },
  });
  if (!response.ok) {
    const payload = (await response.json().catch(() => null)) as {
      detail?: string;
    } | null;
    throw new ReviewApiError(
      payload?.detail ?? "The document review service is unavailable.",
      response.status,
    );
  }
  return (await response.json()) as T;
}

export function listReviewQueue(): Promise<ReviewQueueItem[]> {
  return request<ReviewQueueItem[]>("");
}

export function getReviewDocument(documentId: string): Promise<ReviewDocument> {
  return request<ReviewDocument>(`/${documentId}`);
}

export function getPagePreview(
  documentId: string,
  pageArtifactId: string,
): Promise<{ url: string; expires_in_seconds: number }> {
  return request(`/${documentId}/pages/${pageArtifactId}/preview`);
}

export function decideCandidate(
  documentId: string,
  candidateId: string,
  body: {
    action: ReviewAction;
    expected_candidate_version: number;
    corrected_value?: string;
    corrected_unit?: string;
    reason_code?: string;
    source_verified: boolean;
  },
): Promise<ReviewCandidate> {
  return request(`/${documentId}/candidates/${candidateId}/decisions`, {
    method: "POST",
    headers: { "Idempotency-Key": crypto.randomUUID() },
    body: JSON.stringify(body),
  });
}

export function addManualCandidate(
  documentId: string,
  body: {
    entity_type: string;
    normalized_value: string;
    unit?: string;
    source_page_artifact_id: string;
    source_page_number: number;
    source_verified: boolean;
  },
): Promise<ReviewCandidate> {
  return request(`/${documentId}/candidates`, {
    method: "POST",
    headers: { "Idempotency-Key": crypto.randomUUID() },
    body: JSON.stringify(body),
  });
}

export function finalizeReview(
  documentId: string,
  expectedDocumentVersion: number,
): Promise<ReviewDocument> {
  return request(`/${documentId}/finalize`, {
    method: "POST",
    headers: { "Idempotency-Key": crypto.randomUUID() },
    body: JSON.stringify({
      expected_document_version: expectedDocumentVersion,
    }),
  });
}
